test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java

```diff
@@ -88,4 +99,95 @@ static public void testPassingNullToIsFlattenedThrowsException() throws Throwabl
 			myUnsafe.isFlattened(null);
 		});
 	}
+
+	@Test
+	static public void testValueHeaderSizeOfValueType() throws Throwable {
+		long headerSize = myUnsafe.valueHeaderSize(ValueTypePoint2D.class);
+		if (isCompressedRefsEnabled) {
+			assertEquals(headerSize, 4);
+		} else {
+			assertEquals(headerSize, 8);
+		}
+	}
+
+	@Test
+	static public void testValueHeaderSizeOfClassWithLongField() throws Throwable {
+		long headerSize = myUnsafe.valueHeaderSize(ValueTypeWithLongField.class);
+		if (isCompressedRefsEnabled && !isFlatteningEnabled) {
+			assertEquals(headerSize, 4);
+		} else {
+			// when compressed refs are enabled and the class has an 8 byte field, 4 bytes of padding are added to the class. These padding bytes are included in the headerSize
```

Comment:
> and the class has an 8 byte field, 4 bytes of padding are added

This is not accurate. The 4 bytes padding are added if the class has 8-byte field(s) only.

Due to alignment requirement, 8-byte field needs to start at offset 8, 16, 24, etc. Fields start after the J9Class pointer which is 4 bytes in compressedrefs. If there is an 4-byte field, it will be put at offset 4 (after J9Class pointer) to avoid padding. If no, we will pad 4 bytes to let 8-byte field start at offset 8.

Reply:
oops, those should be fixed now in c8a554b